### PR TITLE
RDM 10737- Introduce Switch for Default Mode of Collection Assertion

### PR DESCRIPTION
**Pull request by sullivangareth**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-10737

### Change description ###

Introduce a switch to the befta framework controlled by an environment variable DEFAULT_COLLECTION_ASSERTION_MODE, so that:

- if the env var is not set, or is set to "ORDERED", the exact current behaviour is preserved.
- if it is set to "UNORDERED", the framework compares all collections as an unordered collection unless specified otherwise  in the test data.  In this default mode of comparison, if the field id(s) to be used as unique identified is/are specified the befta framework uses these.   However, if the id field is not specified, the framework should automatically extract a best candidate from static, expected test data as to which field(s) can be the best one(s) to uniquely identify the elements for matching between expected and actual collections.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/befta/TestAutomationConfig.java [hunk 1]
```diff
@@ -48,6 +49,22 @@ public UserTokenProviderConfig getUserTokenProviderConfig() {
...
+        Ordering returnValue = Ordering.ORDERED;
+
+        try {
+            Ordering defaultCollectionAssertionMode =
+                    Ordering.of(EnvironmentVariableUtils.getOptionalVariable("DEFAULT_COLLECTION_ASSERTION_MODE"));
+            if (defaultCollectionAssertionMode != null) {
+                returnValue = defaultCollectionAssertionMode;
+            }
+        } catch (IllegalArgumentException iae) {
+            returnValue = Ordering.ORDERED;
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/TestAutomationConfig.java:62]: no need for this.

Comment (MSancaktutar) [reply src/main/java/uk/gov/hmcts/befta/TestAutomationConfig.java:62]: But good to generate a warning log.

File: src/main/java/uk/gov/hmcts/befta/data/CollectionVerificationConfig.java [hunk 2]
```diff
@@ -51,4 +56,42 @@ public boolean isDefault() {
...
+                    config.setOrdering(BeftaMain.getConfig().getDefaultCollectionAssertionMode());
+                }
+
+                if (config.getOrdering() == Ordering.UNORDERED) {
+                    String idString = (String) firstMap.get(ELEMENT_ID_FIELD_NAME);
+
+                    if (idString != null) {
+                        config.setElementId(idString);
+                    } else {
+                        config.setElementId(ElementIdFinder.findElementIds(collection));
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/data/CollectionVerificationConfig.java:89]: If element ID finder can't find a good ID candidate, then the test should fail with a test data exception.

File: src/test/java/uk/gov/hmcts/befta/data/CollectionVerificationConfigTest.java [hunk 3]
```diff
@@ -40,4 +59,170 @@ void testToString() {
...
+	void testVerificationConfigFromEmptyCollection() {
+		assertEquals(CollectionVerificationConfig.DEFAULT,
+				CollectionVerificationConfig.getVerificationConfigFrom(unmodifiableCollection(Collections.emptyList())));
+	}
+
+	/**
+	 * Test method for {@link uk.gov.hmcts.befta.data.CollectionVerificationConfig#getVerificationConfigFrom(Collection)} ()}.
+	 */
+	@Test
+	void testVerificationConfigReturnsDefaultFromListCollectionWhereirstElementIsCollectionVerificationConfig() {
```
Comment (MSancaktutar) [inline src/test/java/uk/gov/hmcts/befta/data/CollectionVerificationConfigTest.java:83]: typo in method name.

Comment (sullivangareth) [reply src/test/java/uk/gov/hmcts/befta/data/CollectionVerificationConfigTest.java:83]: Cheers - they are way too long...Need to sort

File: src/test/java/uk/gov/hmcts/befta/util/ElementIdFinderTest.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        Map<String, String> caseRole2 = new HashMap<>();
+        caseRole2.put(CASE_ROLE_KEY, "[CR-2]");
+
+        objectCaseUsers.add(metaData);
+        objectCaseUsers.add(caseRole1);
+        objectCaseUsers.add(caseRole2);
+
+        assertEquals(CASE_ROLE_KEY, ElementIdFinder.findElementIds(objectCaseUsers));
+    }
+}
```
Comment (MSancaktutar) [inline src/test/java/uk/gov/hmcts/befta/util/ElementIdFinderTest.java:216]: eol at eof

File: src/main/java/uk/gov/hmcts/befta/TestAutomationConfig.java (hunk 1 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/befta/TestAutomationConfig.java:62]: Agreed, but better than an empty catch block - Need to work out logging levels in befta as not everything needs to be logged to console by default

File: src/main/java/uk/gov/hmcts/befta/data/CollectionVerificationConfig.java (hunk 2 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/befta/data/CollectionVerificationConfig.java:89]: See line 89 in ElementIDFinder

File: src/test/java/uk/gov/hmcts/befta/data/CollectionVerificationConfigTest.java (hunk 3 above)
Comment (sullivangareth) [reply src/test/java/uk/gov/hmcts/befta/data/CollectionVerificationConfigTest.java]: resolved in https://github.com/hmcts/befta-fw/pull/137/commits/25a5aaeaaf5b1f5144499ab6112c0f2ee82e8329

File: src/test/java/uk/gov/hmcts/befta/util/ElementIdFinderTest.java (hunk 4 above)
Comment (sullivangareth) [reply src/test/java/uk/gov/hmcts/befta/util/ElementIdFinderTest.java]: https://github.com/hmcts/befta-fw/pull/137/commits/25a5aaeaaf5b1f5144499ab6112c0f2ee82e8329

**Review by MSancaktutar — APPROVED** [review]
Great work, congratulations!

**Pull request merged**